tests/jlm/llvm/frontend/llvm/TestLlvmPhiConversion.cpp

```diff
@@ -0,0 +1,256 @@
+/*
+ * Copyright 2024 Håvard Krogstie <krogstie.havard@gmail.com>
+ * See COPYING for terms of redistribution.
+ */
+
+#include <test-registry.hpp>
+#include <test-util.hpp>
+
+#include <jlm/llvm/frontend/LlvmModuleConversion.hpp>
+#include <jlm/llvm/ir/operators/call.hpp>
+#include <jlm/llvm/ir/operators/operators.hpp>
+#include <jlm/llvm/ir/print.hpp>
+#include <jlm/rvsdg/bitstring/constant.hpp>
+
+#include <llvm/IR/BasicBlock.h>
+#include <llvm/IR/IRBuilder.h>
+#include <llvm/IRReader/IRReader.h>
+#include <llvm/Support/SourceMgr.h>
+
+#include <string>
+
+/**
+ * Tests converting instances of ::llvm::PHINode.
+ * Some of the operands have constant values, and some are results from the predecessors.
+ * One of the phi node has its own result as one of its operands.
+ *
+ * The function corresponds to the C code
+ * uint64_t popcount(uint64_t x) {
+ *     uint64_t popcnt = 0;
+ *
+ *     while(true) {
+ *         // 2 phi nodes with 3 possible predecessors here
+ *         // one for x, and one for popcnt
+ *
+ *         if (x == 0)
+ *             break;
+ *         uint64_t rem = x % 2 == 1;
+ *         x >>= 1;
+ *         if (rem) {
+ *             popcnt++;
+ *             continue;
+ *         }
+ *         else {
+ *             continue;
+ *         }
+ *     }
+ *     return popcnt;
+ * }
+ */
+static int
+TestPhiConversion()
+{
+  static const std::string POPCOUNT_PROGRAM = R""""(
+source_filename = "popcount.c"
+target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
+target triple = "x86_64-pc-linux-gnu"
+
+define dso_local i64 @popcount(i64 noundef %0) {
+br label %bb2
+
+bb2:
+; The current value of x
+%3 = phi i64 [%0, %1], [%8, %bb4], [%8, %bb5]
+; The current value of popcount (1 bits seen)
+; Note that this phi is self-referential!
+%4 = phi i64 [0, %1], [%9, %bb4], [%4, %bb5]
+
+; First check if x is 0, and if so jump to the exit
+%5 = icmp eq i64 %3, 0
+br i1 %5, label %bb6, label %bb3
+
+bb3:
+; Check if x % 2 is 1
+%6 = urem i64 %3, 2
+%7 = icmp eq i64 %6, 1
+
+; Also calculate x>>1 now
+%8 = lshr i64 %3, 1
+
+br i1 %7, label %bb4, label %bb5
+
+bb4:
+; Here x was odd, calculate popcount+1
+%9 = add i64 %4, 1
+br label %bb2
+
+bb5:
+; Here x was even, no need to calculate anything
+br label %bb2
+
+bb6:
+ret i64 %4
+}
+)"""";
```

Comment:
This is going to be a nightmare to maintain if the syntax changes. Why not using the LLVM API to setup the test?

Reply:
@phate It could be done with the API as well, but I am not sure it is any more likely that the LLVM IR syntax changes, than that the Builder API changes. I think they are around equally maintainable, but my first instinct was that using Builder is a lot more verbose. That might not actually be true, so I can change it to use the builder.